Improve how-to Search strings content

BillWagner opened the pull request:
Fixes #3331 

Start by combining the programming guide topics for searching using string methods, and searching using regular expressions.

Then, update the samples and descriptions. 

Finally, add LINQ based search samples.

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/how-to/search-strings?branch=pr-en-us-4427)

BillWagner committed:
move to final location

THis commit is meant to put the old content in the correct updated location. Time for a test build.

BillWagner committed:
update existing samples

BillWagner committed:
update text

Also, split code samples into smaller focused snippets.

BillWagner committed:
proofread and review

pkulikov commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,61 @@
+
+## Where does the sought text occur in a string?
+
+The <xref:System.String.IndexOf%2A> and <xref:System.String.LastIndexOf%2A> methods also search for text in strings. These methods return the location of the text being sought. If the text isn't found, they return `-1`. The following example shows a search for the first and last occurrence of the word "methods", and displayed the text between.
```
At the end of the paragraph please replace "...and displayed the text between" with "... and displays the text in between."

pkulikov commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,61 @@
+
+[!code-csharp-interactive[search strings using methods](../../../samples/snippets/csharp/how-to/strings/SearchStrings.cs#1)]
+
+The preceding example demonstrates an important point for using these methods. Searches are **case-sensitive** by default. You use the <xref:System.StringComparison.CurrentCultureIgnoreCase> enum value to specify a case insensitive search. All these methods default to case-sensitive search.
```
There is a slight duplication in this paragraph: 
>Searches are **case-sensitive** by default.

and
>All these methods default to case-sensitive search.

are separated just by one sentence. Was it intentional?

pkulikov commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,61 @@
+ms.author: "wiwagn"
+---
+
+# How to: Search Strings Using String Methods (C# Programming Guide)
```
"Using String Methods" should be omitted.

pkulikov commented on samples/snippets/csharp/how-to/strings/SearchStrings.cs:
```diff
@@ -0,0 +1,157 @@
+            // </Snippet1>
+        }
+
+        private static void SearchByIndex()
```
Curious: why is this method the only one that is private and not called from the `Examples()` method?

BillWagner replied:
Good catch. They were all meant to be private, and all should be called.

pkulikov commented on docs/csharp/programming-guide/strings/index.md:
```diff
@@ -78,7 +78,7 @@ A string is an object of type <xref:System.String> whose value is text. Internal
   
 ## Substrings  
- A substring is any sequence of characters that is contained in a string. Use the <xref:System.String.Substring%2A> method to create a new string from a part of the original string. You can search for one or more occurrences of a substring by using the <xref:System.String.IndexOf%2A> method. Use the <xref:System.String.Replace%2A> method to replace all occurrences of a specified substring with a  ...
+ A substring is any sequence of characters that is contained in a string. Use the <xref:System.String.Substring%2A> method to create a new string from a part of the original string. You can search for one or more occurrences of a substring by using the <xref:System.String.IndexOf%2A> method. Use the <xref:System.String.Replace%2A> method to replace all occurrences of a specified substring with a  ...
```
If the title of How to is updated, then link text: "[How to: Search Strings Using String Methods]" -> "[How to: Search Strings]"

BillWagner commented:
Thanks for the review @pkulikov I've updated this based on your feedback.

BillWagner committed:
respond to feedback.

pkulikov commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,61 @@
+ms.topic: "article"
+helpviewer_keywords: 
+  - "searching strings [C#]"
+  - "strings [C#], searching with String methods"
```
Should we add keywords for regular expressions?

pkulikov commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,61 @@
+# How to: search strings
+
+You have two main strategies that you can use to search for text in strings. Member methods of the <xref:System.String> class search for specific text. Regular expressions search for patterns in text.
+
```
To be consistent with other two "How to" on strings, we might add a note on interactive C# code blocks:
`[!INCLUDE[interactive-note](~/includes/csharp-interactive-note.md)]`

pkulikov commented on samples/snippets/csharp/how-to/strings/SearchStrings.cs:
```diff
@@ -0,0 +1,158 @@
+                }
+            }
+
+            /* Output:
```
Given the fact that code blocks are interactive and one can click **Run** to see output, should we remove comments with expected outputs from snipppets? There are no those in other two "How to" on strings and I thinks that's fine. If such comments can be removed, then all four methods become much shorter and no need to update comments when code is updated occasionally.

pkulikov reviewed (commented):
@BillWagner thank you for making changes and sorry for my second thought review.

BillWagner committed:
respond to feedback

Also, updated the zip file for the samples.

rpetrusha commented on docs/csharp/how-to/concatenate-multiple-strings.md:
```diff
@@ -57,6 +57,7 @@ combines an array of words by adding a space between each word in the array:
 
 [!code-csharp-interactive[string concatenation using LINQ expressions](../../../samples/snippets/csharp/how-to/strings/Concatenate.cs#6)]  
 
+You can try these samples by looking at the code in our [GitHub repository](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings). Or, you can cownload the samples [as a zip file](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings.zip)
```
no comma after "Or"

CamSoper replied:
Is this sentence correct?  I would have written thusly:

You can try these samples by looking at the code in our [GitHub repository](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings), or you can download the samples [as a zip file](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings.zip).

Also, note the typo on "download" and the period at the end.

rpetrusha commented on docs/csharp/how-to/parse-strings-using-split.md:
```diff
@@ -57,7 +57,7 @@ Consecutive instances of any separator produce the empty string in the output ar
 [!code-csharp-interactive[split strings using strings as separators](../../../samples/snippets/csharp/how-to/strings/ParseStringsUsingSplit.cs#5)]
 
-You can try these samples by looking at the code in our [GitHub repository](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings)
+You can try these samples by looking at the code in our [GitHub repository](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings). Or, you can cownload the samples [as a zip file](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings.zip)
```
No comma after "Or"

CamSoper replied:
Same comment as above.

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,66 @@
+
+# How to: search strings
+
+You have two main strategies that you can use to search for text in strings. Member methods of the <xref:System.String> class search for specific text. Regular expressions search for patterns in text.
```
The first sentence is repetitive. Better: "You can use two main strategies..."
"Member methods" --> just "Methods"?

pkulikov replied:
By the way, I've read "member methods" as "non-static methods".

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,66 @@
+
+## Where does the sought text occur in a string?
+
+The <xref:System.String.IndexOf%2A> and <xref:System.String.LastIndexOf%2A> methods also search for text in strings. These methods return the location of the text being sought. If the text isn't found, they return `-1`. The following example shows a search for the first and last occurrence of the word "methods", and displays the text in between.
```
no comma after "methods"

CamSoper replied:
...occurrence of the word "methods", and displays the...

Should the comma be inside the quotes?  It's so hard to know whether to use programmer syntax or English syntax in a technical doc.  😆

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,66 @@
+
+You can try these samples by looking at the code in our [GitHub repository](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings). Or, you can cownload the samples [as a zip file](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings.zip)
+
+## See Also  
```
It would also be good to add a link to \[Best practices for using strings in .NET](https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings)

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,66 @@
+
+ [C# Programming Guide](../programming-guide/index.md)  
+ [Strings](../programming-guide/strings/index.md)  
+ [LINQ and Strings](http://msdn.microsoft.com/library/6c34169f-7a39-436a-98d8-9a7283043942)   
```
Change these links from MSDN to docs.

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,66 @@
+
+The <xref:System.Text.RegularExpressions.Regex?displayProperty=nameWithType> class can be used to search strings. These searches can range in complexity from simple to complicated text patterns.
+
+The following code example searches for the word "code" in a sentence, ignoring case. The static method <xref:System.Text.RegularExpressions.Regex.IsMatch%2A?displayProperty=nameWithType> performs the search. You give it the string to search and a search pattern. In this case, a third argument specifies case-insensitive search. For more information, see <xref:System.Text.RegularExpressions.RegexO ...
```
Rather than present an example that can better be done with a string method (so we're illustrating a worst practice), I'd do one that actually uses a pattern. Possibly a simple search, such as a one that uses the pattern `the(ir)?` to match either "the" or "their"

pkulikov replied:
Or:
- search for a noun in both singular and plural form
- search for both '2' and 'two'

rpetrusha replied:
nit: if you keep the existing example, you should change the description to note that it search for the first occurrence of the word "code" in a sentence.

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,66 @@
+
+## Does a string follow a pattern?
+
+The following code uses regular expressions to validate the format of each string in an array. The validation requires that each string have the form of a telephone number in which three groups of digits are separated by dashes, the first two groups contain three digits, and the third group contains four digits. The search pattern uses the regular expression `^\\d{3}-\\d{3}-\\d{4}$`. For more inf ...
```
In the regex section of the .NET standard docs, I've generally tried to add a table that breaks down each regex so that they an be better understood. I'd recommend doing that here. (e.g., ^ -- begin at the beginning of the string, etc.)

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,66 @@
+
+This single search pattern matches many valid strings. Regular expressions are better to search for or validate against a pattern, rather than a single text string.
+
+You can try these samples by looking at the code in our [GitHub repository](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings). Or, you can cownload the samples [as a zip file](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/how-to/strings.zip)
```
No comma after Or

CamSoper replied:
Same comment as before.

rpetrusha reviewed (changes requested):
This looks good, @Billwagner. I've left a number of suggestions and comments.

CamSoper reviewed (changes requested):
Some nits.

BillWagner committed:
feedback, part 1

Other feedback involves new samples, so that will be a separate commit.

pkulikov commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -54,13 +54,14 @@ The following code uses regular expressions to validate the format of each strin
- [Regular Expression Language - Quick Reference](http://msdn.microsoft.com/library/930653a6-95d2-4697-9d5a-52d11bb6fd4c)   
+ [.NET Framework Regular Expressions](../../standard/base-types/regular-expressions)   
+ [Regular Expression Language - Quick Reference](../../standard/base-types/regular-expression-language-quick-reference)   
+ [Best practices for using strings in .NET](../../standard/base-types/best-practices-strings)  
```
@BillWagner I'm afraid all those inside-repo links require `.md` extension suffix.

BillWagner replied:
thanks. On it.

BillWagner committed:
fix links

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,67 @@
+
+## Does a string contain text?
+
+The <xref:System.String.Contains%2A?displayProperty=nameWithType>, <xref:System.String.StartsWith%2A?displayProperty=nameWithType> and <xref:System.String.EndsWith%2A?displayProperty=nameWithType> methods search a string for specific text. The following example shows each of these methods, and a variation that uses a case insensitive search:
```
nit: either no comma before "and", or change the last phrase to read "and a variation uses..."

rpetrusha commented on docs/csharp/how-to/search-strings.md:
```diff
@@ -0,0 +1,67 @@
+
+[!code-csharp-interactive[search strings using methods](../../../samples/snippets/csharp/how-to/strings/SearchStrings.cs#1)]
+
+The preceding example demonstrates an important point for using these methods. Searches are **case-sensitive** by default. You use the <xref:System.StringComparison.CurrentCultureIgnoreCase> enum value to specify a case insensitive search.
```
nit: it's best for this to be fully qualified (?displayProperty=nameWithType)

rpetrusha reviewed (approved):
I left a couple of nits that you can decide whether you want to address, @BillWagner. If you want to address any of them, and once you fix the broken links indicated by @pkulikov, this is ready to merge.

BillWagner committed:
update regex sample

Three changes in this commit:
1. update the first Regex sample
2. Add a table for each RegEx to describe its format.
3. rebuild the downloadable sample.

BillWagner committed:
fix two nits
